core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -545,42 +697,101 @@ class ReplicaManager(val config: KafkaConfig,
     Option(allPartitions.get(topicPartition)).getOrElse(HostedPartition.None)
   }
 
-  def isAddingReplica(topicPartition: TopicPartition, replicaId: Int): Boolean = {
+  def isAddingReplica(topicPartition: TopicPartition, replicaId: Int, includeFenced: Boolean): Boolean = {
     getPartition(topicPartition) match {
-      case Online(partition) => partition.isAddingReplica(replicaId)
+      case HostedPartition.Online(partition) => partition.isAddingReplica(replicaId)
+      case HostedPartition.Fenced(metadata) => includeFenced && metadata.partition.isAddingReplica(replicaId)
       case _ => false
     }
   }
 
   // Visible for testing
-  def createPartition(topicPartition: TopicPartition): Partition = {
+  def createPartition(topicPartition: TopicPartition, fenced: Boolean = false,
+                      metadataPartitionForFence: Option[MetadataPartition] = None,
+                      wasNewForFence: Option[Boolean] = None,
+                      mostRecentMetadataOffsetForFence: Option[Long] = None): Partition = {
     val partition = Partition(topicPartition, time, this)
-    allPartitions.put(topicPartition, HostedPartition.Online(partition))
+    allPartitions.put(topicPartition,
+      if (fenced) {
+        if (metadataPartitionForFence.isEmpty || wasNewForFence.isEmpty || mostRecentMetadataOffsetForFence.isEmpty)
+          throw new IllegalArgumentException(s"Must supply all metadata when creating a fenced partition: $topicPartition")
+        HostedPartition.Fenced(FencedPartitionMetadata(partition,
+          metadataPartitionForFence.get, wasNewForFence.get, mostRecentMetadataOffsetForFence.get))
+      } else {
+        HostedPartition.Online(partition)
+      })
     partition
   }
 
   def nonOfflinePartition(topicPartition: TopicPartition): Option[Partition] = {
     getPartition(topicPartition) match {
       case HostedPartition.Online(partition) => Some(partition)
+      case HostedPartition.Fenced(metadata) => Some(metadata.partition)
       case HostedPartition.None | HostedPartition.Offline => None
     }
   }
 
+  // not sure if any of the 9 invocations of the above nonOfflinePartition() should be diverted here
+  def onlinePartition(topicPartition: TopicPartition): Option[Partition] = {
+    getPartition(topicPartition) match {
+      case HostedPartition.Online(partition) => Some(partition)
+      case _ => None
+    }
+  }
+
+  def fencedPartition(topicPartition: TopicPartition): Option[Partition] = {
```

Comment:
nit: the only caller of this does not need access to the `Partition` object. Could we have an `isFenced(TopicPartition)` method instead?